### PR TITLE
Fixed igloos being labeled as haystacks

### LINKED ISSUE #13936: Igloos in TD are named Haystacks.
On the WINTER & SNOW tilesets the civilian structures V12 & V13 display as Igloos but are still named "Haystacks".

openra-2017-08-29t155230768z

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/13936

### CODE REVIEW

File: mods/cnc/rules/defaults.yaml [hunk 1]
```diff
@@ -813,6 +813,26 @@
 	RenderSprites:
 		Palette: terrain
 
+^CivHaystack:
```
Comment (pchote) [inline mods/cnc/rules/defaults.yaml]: How about `^CivHaystackOrIgloo` to make it clear?

File: mods/cnc/rules/defaults.yaml [hunk 2]
```diff
@@ -813,6 +813,26 @@
...
+		Name: Igloo
+		RequiresCondition: winter
+		GenericVisibility: Enemy, Neutral
+		ShowOwnerRow: false
+	GrantConditionOnTileSet@SUMMER:
+		Condition: summer
+		TileSets: TEMPERAT
+	Tooltip@SUMMER:
+		Name: Haystacks
+		RequiresCondition: summer
```
Comment (pchote) [inline mods/cnc/rules/defaults.yaml]: `RequiresCondition: !winter` and `GrantConditionOnTileSet@SUMMER` can be removed.

File: mods/cnc/rules/defaults.yaml [hunk 3]
```diff
@@ -831,6 +851,26 @@
 	MapEditorData:
 		Categories: Husk
 
+^CivHaystackHusk:
```
Comment (pchote) [inline mods/cnc/rules/defaults.yaml]: Likewise here.

File: mods/cnc/rules/defaults.yaml [hunk 4]
```diff
@@ -831,6 +851,26 @@
...
+		Name: Igloo (Destroyed)
+		RequiresCondition: winter
+		GenericVisibility: Enemy, Neutral
+		ShowOwnerRow: false
+	GrantConditionOnTileSet@SUMMER:
+		Condition: summer
+		TileSets: TEMPERAT
+	Tooltip@SUMMER:
+		Name: Haystacks (Destroyed)
+		RequiresCondition: summer
```
Comment (pchote) [inline mods/cnc/rules/defaults.yaml]: Likewise here.

**Comment by Mailaender** [comment]
Updated.

File: mods/cnc/rules/campaign-tooltips.yaml [hunk 5]
```diff
@@ -42,6 +42,14 @@
 		GenericVisibility: Enemy, Neutral
 		ShowOwnerRow: false
 
+^CivHaystack:
```
Comment (abcdefg30) [inline mods/cnc/rules/campaign-tooltips.yaml]: Doesn't this need to be `^CivHaystackOrIgloo` as well now?

Comment (Mailaender) [reply mods/cnc/rules/campaign-tooltips.yaml]: Yes.

File: mods/cnc/rules/defaults.yaml [hunk 6]
```diff
@@ -813,6 +813,23 @@
...
+		ExcludeTilesets: DESERT
+	-Tooltip:
+	GrantConditionOnTileSet@WINTER:
+		Condition: winter
+		TileSets: SNOW, WINTER
+	Tooltip@WINTER:
+		Name: Igloo
+		RequiresCondition: winter
+		GenericVisibility: Enemy, Neutral
+		ShowOwnerRow: false
```
Comment (reaperrr) [inline mods/cnc/rules/defaults.yaml:828]: `GenericVisibility` and `ShowOwnerRow` values here seem to be identical to the campaign overrides, while the SUMMER tooltip below doesn't define overrides for them here, only in `campaign-tooltips.yaml`.
Is that intentional?

File: mods/cnc/rules/defaults.yaml [hunk 7]
```diff
@@ -831,6 +848,23 @@
...
+		ExcludeTilesets: DESERT
+	-Tooltip:
+	GrantConditionOnTileSet@WINTER:
+		Condition: winter
+		TileSets: SNOW, WINTER
+	Tooltip@WINTER:
+		Name: Igloo (Destroyed)
+		RequiresCondition: winter
+		GenericVisibility: Enemy, Neutral
+		ShowOwnerRow: false
```
Comment (reaperrr) [inline mods/cnc/rules/defaults.yaml:865]: Same question here, except this seems to be missing the `campaign-tooltips.yaml` override, which means the summer haystacks will use the internal Tooltip defaults. Is that intended?

**Comment by Mailaender** [comment]
Updated the tooltips to match the fields and added a fix for editor tooltips.

File: OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs [hunk 8]
```diff
@@ -79,8 +79,8 @@ public EditorActorPreview(WorldRenderer worldRenderer, string id, ActorReference
 
 			Footprint = ios?.OccupiedCells(Info, location, subCell) ?? new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } };
 
-			tooltip = Info.TraitInfos<EditorOnlyTooltipInfo>().FirstOrDefault(info => info.EnabledByDefault) as TooltipInfoBase
-				?? Info.TraitInfos<TooltipInfo>().FirstOrDefault(info => info.EnabledByDefault);
+			tooltip = Info.TraitInfos<EditorOnlyTooltipInfo>().FirstEnabledTraitOrDefault() as TooltipInfoBase
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs]: I'm not sure if this works properly. We have Info classes here, but FirstEnabled*Trait*OrDefault only works on ConditionalTraits I think? (The function uses `IDisabledTrait` which is only implemented by `ConditionalTrait`.)

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs]: grafik

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs]: The debugger shows me that the Igloo TooltipInfo was always the first of the two, and it was always chosen as well.

**Review by abcdefg30 — CHANGES_REQUESTED** [review]
Regardless of the editor thing, can we adjust RA as well?
grafik

**Comment by Mailaender** [comment]
Updated Red Alert and removed the non-functional patch for the editor.

**Review by abcdefg30 — APPROVED** [review]

**Comment by Smittytron** [comment]
I'm getting igloo in game but haystacks in editor, is this intended?

**Comment by Mailaender** [comment]
No, not intended, but I also don't have a solution for it as it is part of a bigger problem
https://github.com/OpenRA/OpenRA/blob/dcaa658678c398b600a7b591a4d4cf796ea50493/OpenRA.Mods.Common/Traits/Conditions/ConditionalTrait.cs#L25-L28

**Review by Smittytron — APPROVED** [review]
> No, not intended, but I also don't have a solution for it as it is part of a bigger problem

> abcdefg30: that is better suited for a follow-up, so you can go ahead @Smitty

OK, we'll leave it for a follow-up then.

**Comment by Smittytron** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b481eeddb2ac69faa9b07326da1beaddea03f7e3

**Pull request merged**